srml/support/src/storage/hashed/generator.rs

```diff
@@ -295,16 +341,61 @@ pub trait EnumerableStorageMap<K: codec::Codec, V: codec::Codec>: StorageMap<K,
 pub trait AppendableStorageMap<K: codec::Codec, V: codec::Codec>: StorageMap<K, V> {
 	/// Append the given items to the value in the storage.
 	///
-	/// `T` is required to implement `codec::EncodeAppend`.
+	/// `V` is required to implement `codec::EncodeAppend`.
 	fn append<S: HashedStorage<Self::Hasher>, I: codec::Encode>(
-		key : &K, items: &[I], storage: &mut S
-	) -> Result<(), &'static str> where V: codec::EncodeAppend<Item=I> {
+		key : &K,
+		items: &[I],
+		storage: &mut S
+	) -> Result<(), &'static str>
+		where V: codec::EncodeAppend<Item=I>
+	{
 		let k = Self::key_for(key);
 		let new_val = <V as codec::EncodeAppend>::append(
-			storage.get_raw(&k[..]).unwrap_or_default(),
+			storage.get_raw(&k[..]).unwrap_or_else(|| {
+				// otherwise, try and read a proper __provided__ default.
+				Self::Default::default().map(|v| v.encode())
+					// or just use the default value.
+					.unwrap_or_default()
+			}),
 			items,
 		).map_err(|_| "Could not append given item")?;
 		storage.put_raw(&k[..], &new_val);
 		Ok(())
 	}
+
+	/// Safely append the given items to the value in the storage. If a codec error occurs, then the
+	/// old (presumably corrupt) value is replaced with the given `items`.
+	///
+	/// `T` is required to implement `codec::EncodeAppend`.
+	fn append_or_put<'a, S, I>(
```

Comment:
```suggestion
	fn append_or_insert<'a, S, I>(
```

I would stick to the terminology of the map.